### ISSUE TITLE
Accessing array item at any index always produces a defined value and it shouldn't

### ISSUE THREAD

**Issue by fingerartur** [google/closure-compiler#4026]
Accessing an array of numbers on any index always gives a value with type `number`. It should give a value with type `number|undefined`. The same is true of any type, not just number.

```js
/**
 * @type {Array<number>}
*/
const values = [1, 2, 3, 4, 5]
/**
* @type {number}
*/
const x = values[1000] // This should raise error. x is obviously undefined, but the compiler thinks it is a number instead
```

Compiler Version: v20221102

Build command:

```
java -jar ./scripts/closureCompiler.jar \
  --entry_point=./src/js/index.js \
  --js=./src/**.js \
  --dependency_mode=PRUNE \
  --warning_level=VERBOSE \
  --js_output_file=./dist/bundle.js \
  --module_resolution=WEBPACK \
  --compilation_level=ADVANCED \
  --jscomp_error=checkDebuggerStatement \
  --jscomp_error=unusedLocalVariables \
  --jscomp_error=reportUnknownTypes \
  --jscomp_error=strictCheckTypes;
```

**Comment by fingerartur** [comment]
Actually I might have jumped the gun on this one. I analysed this a little bit more and Typescript behaves the same as Google Closure in this case. The truth is, the current behavior of the compiler is probably more convenient to programmers even though it may not be 100% type-safe.

For inspiration: [Typescript actually has a config flag](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-1.html#checked-indexed-accesses---nouncheckedindexedaccess) to switch between 100% type-safe behavior and more convenient behavior and this flag is off by default.

In conclusion: We can close this issue and keep the current behavior.
